File: Sources/AppModel.swift
//
// Copyright (c) 2026 Nikita Denin <nik@brite-apps.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import AppKit
import Foundation
import SwiftUI

@MainActor
final class AppModel: ObservableObject {
    @Published private(set) var externalCandidates: [VolumeCandidate] = []
    @Published var selectedExternalRoot: URL? {
        didSet {
            persistSelectedRoot(selectedExternalRoot, key: .externalRoot)
        }
    }
    @Published var selectedSyncRoot: URL? {
        didSet {
            persistSelectedRoot(selectedSyncRoot, key: .syncRoot)
        }
    }
    @Published private(set) var rows: [SaveComparison] = []
    @Published private(set) var catalog: GameCatalog = .empty
    @Published var selectedRowIDs = Set<SaveComparison.ID>()
    @Published var useSerialStationAPI = false
    @Published var backupsEnabled: Bool {
        didSet {
            UserDefaults.standard.set(backupsEnabled, forKey: Self.backupsEnabledKey)
        }
    }
    @Published var statusMessage = String(localized: "Choose a PSP storage root and a sync root.")
    @Published var isWorking = false
    @Published var alert: AppAlert?
    @Published private(set) var backups: [BackupArchive] = []
    @Published var selectedBackupID: BackupArchive.ID?

    private let scanner = SaveScanner()
    private let syncEngine = SyncEngine()
    private let backupStore = BackupStore()
    private let directoryBookmarkStore = DirectoryBookmarkStore()
    private let catalogStore = CatalogStore()
    private let serialStationClient = SerialStationClient()
    private let serialStationCacheStore = SerialStationCacheStore()
    private var serialStationCache: [String: GameMetadata] = [:]
    private var accessedDirectoryPaths = Set<String>()
    private var promptedExternalRootAccessPaths = Set<String>()
    private static let backupsEnabledKey = "backupsEnabled"

    init() {
        backupsEnabled = UserDefaults.standard.object(forKey: Self.backupsEnabledKey) as? Bool ?? true
        loadSerialStationCache()
        restoreSelectedRoots()
        refreshBackups()
        Task {
            await loadCatalog()
            await refreshVolumes()
        }
    }

    deinit {
        for path in accessedDirectoryPaths {
            URL(fileURLWithPath: path).stopAccessingSecurityScopedResource()
        }
    }

    func refreshVolumes() async {
        externalCandidates = scanner.discoverVolumeCandidates()
        if selectedExternalRoot == nil, let candidate = externalCandidates.first {
            await requestAccessToDetectedExternalRoot(candidate.rootURL)
        }
        guard selectedExternalRoot != nil else {
            rows = []
            if externalCandidates.isEmpty {
                statusMessage = String(localized: "No PSP storage selected.")
            }
            return
        }
        await scan()
    }

    func loadCatalog() async {
        do {
            catalog = try catalogStore.loadCatalog()
        } catch {
            alert = AppAlert(title: String(localized: "Catalog could not be loaded"), message: error.localizedDescription)
        }
    }

    func importCatalog(from url: URL) async {
        do {
            catalog = try catalogStore.importCatalog(from: url)
            await scan()
            statusMessage = String(localized: "Imported \(catalog.games.count) catalog entries.")
        } catch {
            alert = AppAlert(title: String(localized: "Catalog import failed"), message: error.localizedDescription)
        }
    }

    func selectExternalRoot() {
        if let url = DirectoryPicker.pick(
            title: String(localized: "Select PSP storage root"),
            message: String(localized: "Choose the top-level PSP volume or memory card folder that contains PSP/SAVEDATA."),
            directoryURL: selectedExternalRoot ?? externalCandidates.first?.rootURL
        ) {
            setExternalRoot(url)
            Task {
                await scan()
            }
        }
    }

    func selectSyncRoot() {
        if let url = DirectoryPicker.pick(
            title: String(localized: "Select sync root"),
            message: String(localized: "Choose the top-level folder that contains or will contain PSP/SAVEDATA.")
        ) {
            setSyncRoot(url)
            Task {
                await scan()
            }
        }
    }

    func setExternalRoot(_ url: URL?) {
        selectedExternalRoot = url
    }

    func setSyncRoot(_ url: URL?) {
        selectedSyncRoot = url
    }

    func scan() async {
        guard let externalRoot = selectedExternalRoot else {
            rows = []
            statusMessage = String(localized: "No PSP storage selected.")
            return
        }
        let syncRoot = selectedSyncRoot
        let catalog = catalog

        isWorking = true
        defer {
            isWorking = false
        }

        do {
            var (pspSaves, syncSaves) = try await Task.detached(priority: .userInitiated) {
                let scanner = SaveScanner()
                let pspSaves = try scanner.scanPSPSaves(storageRoot: externalRoot, catalog: catalog)
                let syncSaves = try syncRoot.map { try scanner.scanSyncSaves(syncRoot: $0, catalog: catalog) } ?? []
                return (pspSaves, syncSaves)
            }.value

            var metadata = serialStationCache
            if useSerialStationAPI {
                metadata = await fetchSerialStationMetadata(for: pspSaves + syncSaves)
            }
            if !metadata.isEmpty {
                pspSaves = pspSaves.map { $0.enriched(with: metadata[$0.gameID] ?? $0.game) }
                syncSaves = syncSaves.map { $0.enriched(with: metadata[$0.gameID] ?? $0.game) }
            }

            rows = syncEngine.compare(pspSaves: pspSaves, syncSaves: syncSaves)
            selectedRowIDs = Set(rows.filter { $0.state != .same }.map(\.id))
            statusMessage = String(localized: "\(rows.count) save folders found.")
        } catch {
            rows = []
            statusMessage = String(localized: "Scan failed.")
            alert = AppAlert(title: String(localized: "Scan failed"), message: error.localizedDescription)
        }
    }

    func syncSelected() async {
        guard let syncRoot = selectedSyncRoot else {
            alert = AppAlert(title: String(localized: "Sync root required"), message: String(localized: "Select a sync root before syncing."))
            return
        }

        guard let externalRoot = selectedExternalRoot else {
            alert = AppAlert(title: String(localized: "PSP storage required"), message: String(localized: "Select PSP storage before syncing."))
            return
        }

        let selectedRows = rows.filter { selectedRowIDs.contains($0.id) }
        let willWriteChanges = selectedRows.contains { $0.state != .same }
        let shouldCreateBackup = backupsEnabled && willWriteChanges

        isWorking = true
        defer {
            isWorking = false
        }

        do {
            if shouldCreateBackup {
                let backupURL = try await Task.detached(priority: .userInitiated) {
                    try BackupStore().createBackup(of: syncRoot)
                }.value
                refreshBackups()
                statusMessage = String(localized: "Created backup \(backupURL.lastPathComponent).")
            }
            let syncedCount = try await Task.detached(priority: .userInitiated) {
                try SyncEngine().syncLatest(rows: selectedRows, syncRoot: syncRoot, pspRoot: externalRoot)
            }.value
            await scan()
            statusMessage = syncedCount == 0 ? String(localized: "Everything is already in sync.") : String(localized: "Synced \(syncedCount) save folders.")
        } catch {
            alert = AppAlert(title: String(localized: "Sync failed"), message: error.localizedDescription)
        }
    }

    func deletePSPSave(_ row: SaveComparison) async {
        guard let save = row.psp else {
            return
        }
        await deleteSave(
            at: save.rootURL,
            title: String(localized: "Delete from PSP storage?"),
            message: String(localized: "This will permanently delete \(row.displayTitle) from PSP storage."),
            status: String(localized: "Deleted \(row.displayTitle) from PSP storage.")
        )
    }

    func deleteSyncSave(_ row: SaveComparison) async {
        guard let save = row.sync else {
            return
        }
        await deleteSave(
            at: save.rootURL,
            title: String(localized: "Delete from sync root?"),
            message: String(localized: "This will permanently delete \(row.displayTitle) from the sync root."),
            status: String(localized: "Deleted \(row.displayTitle) from the sync root.")
        )
    }

    func deleteBothSaves(_ row: SaveComparison) async {
        guard let pspSave = row.psp, let syncSave = row.sync else {
            return
        }
        await deleteSaves(
            urls: [pspSave.rootURL, syncSave.rootURL],
            title: String(localized: "Delete both copies?"),
            message: String(localized: "This will permanently delete \(row.displayTitle) from PSP storage and the sync root."),
            status: String(localized: "Deleted \(row.displayTitle) from both sides.")
        )
    }

    func restoreSelectedBackup() async {
        guard let syncRoot = selectedSyncRoot else {
            alert = AppAlert(title: String(localized: "Sync root required"), message: String(localized: "Select a sync root before restoring a backup."))
            return
        }

        guard let backup = backups.first(where: { $0.id == selectedBackupID }) else {
            alert = AppAlert(title: String(localized: "Backup required"), message: String(localized: "Select a backup before restoring."))
            return
        }
        let backupURL = backup.url

        let confirmed = ConfirmationDialog.ask(
            title: String(localized: "Restore backup?"),
            message: String(localized: "This will replace the current sync root contents with \(backupURL.lastPathComponent)."),
            confirmTitle: String(localized: "Restore")
        )
        guard confirmed else {
            return
        }

        isWorking = true
        defer {
            isWorking = false
        }

        do {
            if backupsEnabled {
                let currentBackupURL = try backupStore.createBackup(of: syncRoot, preserving: backupURL)
                refreshBackups()
                statusMessage = String(localized: "Created backup \(currentBackupURL.lastPathComponent).")
            }
            try backupStore.restoreBackup(from: backupURL, to: syncRoot)
            await scan()
            refreshBackups()
            statusMessage = String(localized: "Restored \(backupURL.lastPathComponent).")
        } catch {
            alert = AppAlert(title: String(localized: "Restore failed"), message: error.localizedDescription)
        }
    }

    func refreshBackups() {
        do {
            backups = try backupStore.listBackups()
            if selectedBackupID == nil || !backups.contains(where: { $0.id == selectedBackupID }) {
                selectedBackupID = backups.first?.id
            }
        } catch {
            backups = []
            selectedBackupID = nil
            alert = AppAlert(title: String(localized: "Backups could not be loaded"), message: error.localizedDescription)
        }
    }

    private func fetchSerialStationMetadata(for saves: [SaveGame]) async -> [String: GameMetadata] {
        let ids = Array(Set(saves.map(\.gameID))).filter { GameIDParser.parse(from: $0) != nil }
        var result = serialStationCache
        let missingIDs = ids.filter { result[$0] == nil }

        if missingIDs.isEmpty {
            return result
        }

        statusMessage = String(localized: "Looking up \(missingIDs.count) titles on SerialStation...")

        let client = serialStationClient

        await withTaskGroup(of: (String, GameMetadata?).self) { group in
            for id in missingIDs {
                group.addTask {
                    let metadata = try? await client.metadata(for: id)
                    return (id, metadata)
                }
            }

            for await (id, metadata) in group {
                if let metadata {
                    result[id] = metadata
                }
            }
        }

        serialStationCache = result
        saveSerialStationCache()
        return result
    }

    private func deleteSave(at url: URL, title: String, message: String, status: String) async {
        await deleteSaves(urls: [url], title: title, message: message, status: status)
    }

    private func deleteSaves(urls: [URL], title: String, message: String, status: String) async {
        let confirmed = ConfirmationDialog.ask(
            title: title,
            message: message,
            confirmTitle: String(localized: "Delete")
        )
        guard confirmed else {
            return
        }

        isWorking = true
        defer {
            isWorking = false
        }

        do {
            for url in urls {
                try FileManager.default.removeItem(at: url)
            }
            await scan()
            statusMessage = status
        } catch {
            alert = AppAlert(title: String(localized: "Delete failed"), message: error.localizedDescription)
        }
    }

    private func loadSerialStationCache() {
        serialStationCache = (try? serialStationCacheStore.load()) ?? [:]
    }

    private func saveSerialStationCache() {
        try? serialStationCacheStore.save(serialStationCache)
    }

    private func requestAccessToDetectedExternalRoot(_ url: URL, forcePrompt: Bool = false) async {
        let standardizedURL = url.standardizedFileURL
        let path = standardizedURL.path
        if !forcePrompt, promptedExternalRootAccessPaths.contains(path) {
            statusMessage = String(localized: "PSP storage detected. Choose it to grant access.")
            return
        }

        promptedExternalRootAccessPaths.insert(path)
        guard let selectedURL = DirectoryPicker.pick(
            title: String(localized: "Select PSP storage root"),
            message: String(localized: "Choose the top-level PSP volume or memory card folder that contains PSP/SAVEDATA."),
            directoryURL: standardizedURL
        ) else {
            statusMessage = String(localized: "PSP storage detected. Choose it to grant access.")
            return
        }

        setExternalRoot(selectedURL)
    }

    private func restoreSelectedRoots() {
        do {
            if let externalRoot = try directoryBookmarkStore.loadURL(for: .externalRoot) {
                selectedExternalRoot = externalRoot
            }
            if let syncRoot = try directoryBookmarkStore.loadURL(for: .syncRoot) {
                selectedSyncRoot = syncRoot
            }
        } catch {
            alert = AppAlert(title: String(localized: "Saved folders could not be restored"), message: error.localizedDescription)
        }
    }

    private func persistSelectedRoot(_ url: URL?, key: DirectoryBookmarkStore.Key) {
        guard let url else {
            return
        }

        let standardizedURL = url.standardizedFileURL
        let path = standardizedURL.path
        if !accessedDirectoryPaths.contains(path), standardizedURL.startAccessingSecurityScopedResource() {
            accessedDirectoryPaths.insert(path)
        }

        do {
            try directoryBookmarkStore.save(standardizedURL, for: key)
        } catch {
            alert = AppAlert(title: String(localized: "Folder selection could not be saved"), message: error.localizedDescription)
        }
    }
}

struct AppAlert: Identifiable {
    let id = UUID()
    let title: String
    let message: String
}

enum DirectoryPicker {
    static func pick(title: String, message: String, directoryURL: URL? = nil) -> URL? {
        let panel = NSOpenPanel()
        panel.title = title
        panel.message = message
        panel.canChooseFiles = false
        panel.canChooseDirectories = true
        panel.allowsMultipleSelection = false
        panel.canCreateDirectories = true
        panel.directoryURL = directoryURL
        return panel.runModal() == .OK ? panel.url : nil
    }
}

enum ConfirmationDialog {
    static func ask(title: String, message: String, confirmTitle: String) -> Bool {
        let alert = NSAlert()
        alert.messageText = title
        alert.informativeText = message
        alert.alertStyle = .warning
        alert.addButton(withTitle: confirmTitle)
        alert.addButton(withTitle: String(localized: "Cancel"))
        return alert.runModal() == .alertFirstButtonReturn
    }
}

File: wpspsync_flutter/lib/services/app_model.dart
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:file_picker/file_picker.dart';
import 'package:path/path.dart' as p;
import '../models/models.dart';
import 'scanner_service.dart';
import 'sync_service.dart';
import 'backup_service.dart';
import 'catalog_service.dart';
import 'serial_station_service.dart';
import 'serial_station_cache_service.dart';

class AppModel extends ChangeNotifier {
  // Observable state
  List<VolumeCandidate> externalCandidates = [];
  Directory? selectedExternalRoot;
  Directory? selectedSyncRoot;
  List<SaveComparison> rows = [];
  GameCatalog catalog = GameCatalog.empty;
  Set<String> selectedRowIDs = {};
  bool useSerialStationAPI = false;
  bool backupsEnabled = true;
  String statusMessage = "Choose a PSP storage root and a sync root.";
  bool isWorking = false;
  bool isSyncing = false;
  List<BackupArchive> backups = [];
  String? selectedBackupId;

  // In-memory SerialStation cache (also persisted to disk)
  Map<String, GameMetadata> _serialStationCache = {};

  // Services
  final ScannerService _scanner = ScannerService();
  final SyncService _syncEngine = SyncService();
  final BackupService _backupStore = BackupService();
  final CatalogService _catalogStore = CatalogService();
  final SerialStationService _serialStation = SerialStationService();
  final SerialStationCacheService _serialStationCache_ = SerialStationCacheService();

  // Preference keys
  static const _externalRootKey = "externalRoot";
  static const _syncRootKey = "syncRoot";
  static const _backupsEnabledKey = "backupsEnabled";
  static const _useSerialStationKey = "useSerialStationAPI";

  AppModel() {
    _initialize();
  }

  Future<void> _initialize() async {
    final prefs = await SharedPreferences.getInstance();
    backupsEnabled = prefs.getBool(_backupsEnabledKey) ?? true;
    useSerialStationAPI = prefs.getBool(_useSerialStationKey) ?? false;

    // Load the on-disk SerialStation cache first
    _serialStationCache = await _serialStationCache_.load();

    await _restoreSelectedRoots();
    await refreshBackups();
    await loadCatalog();
    await refreshVolumes();
  }

  // Refreshes detected external volumes and auto-selects if needed
  Future<void> refreshVolumes() async {
    externalCandidates = await _scanner.discoverVolumeCandidates();

    if (selectedExternalRoot == null && externalCandidates.isNotEmpty) {
      selectedExternalRoot = Directory(externalCandidates.first.rootUri.toFilePath());
    }

    if (selectedExternalRoot == null) {
      rows = [];
      statusMessage = externalCandidates.isEmpty
          ? "No PSP storage selected."
          : "PSP storage detected. Choose it to grant access.";
    } else {
      await scan();
    }
    notifyListeners();
  }

  // Loads the game catalog from disk or bundled asset
  Future<void> loadCatalog() async {
    try {
      catalog = await _catalogStore.loadCatalog();
      notifyListeners();
    } catch (e) {
      statusMessage = "Catalog could not be loaded: $e";
      notifyListeners();
    }
  }

  // Full scan: discovers saves on both sides and compares them
  Future<void> scan() async {
    if (selectedExternalRoot == null) {
      rows = [];
      statusMessage = "No PSP storage selected.";
      notifyListeners();
      return;
    }

    isWorking = true;
    notifyListeners();

    try {
      var pspSaves = await _scanner.scanPSPSaves(selectedExternalRoot!, catalog);
      var syncSaves = selectedSyncRoot != null
          ? await _scanner.scanSyncSaves(selectedSyncRoot!, catalog)
          : <SaveGame>[];

      // Enrich with SerialStation metadata (uses disk cache, fetches only missing IDs)
      if (useSerialStationAPI) {
        final metadata = await _fetchSerialStationMetadata(pspSaves + syncSaves);
        if (metadata.isNotEmpty) {
          pspSaves = pspSaves.map((s) => s.enriched(metadata[s.gameId])).toList();
          syncSaves = syncSaves.map((s) => s.enriched(metadata[s.gameId])).toList();
        }
      } else if (_serialStationCache.isNotEmpty) {
        // Always apply cached metadata even when API toggle is off
        pspSaves = pspSaves.map((s) => s.enriched(_serialStationCache[s.gameId])).toList();
        syncSaves = syncSaves.map((s) => s.enriched(_serialStationCache[s.gameId])).toList();
      }

      rows = _syncEngine.compare(pspSaves: pspSaves, syncSaves: syncSaves);

      // Auto-select all rows that need action
      selectedRowIDs = rows
          .where((r) => r.state != SaveState.same)
          .map((r) => r.id)
          .toSet();

      statusMessage = "${rows.length} save folders found.";
    } catch (e) {
      rows = [];
      statusMessage = "Scan failed: $e";
    } finally {
      isWorking = false;
      notifyListeners();
    }
  }

  // Syncs all selected rows, creating a backup first if enabled
  Future<void> syncSelected() async {
    if (selectedSyncRoot == null || selectedExternalRoot == null) return;

    final selectedRows = rows.where((r) => selectedRowIDs.contains(r.id)).toList();
    final willWriteChanges = selectedRows.any((r) => r.state != SaveState.same);
    final shouldCreateBackup = backupsEnabled && willWriteChanges;

    isWorking = true;
    isSyncing = true;
    notifyListeners();

    try {
      if (shouldCreateBackup) {
        final backupFile = await _backupStore.createBackup(selectedSyncRoot!);
        await refreshBackups();
        statusMessage = "Created backup ${p.basename(backupFile.path)}.";
        notifyListeners();
      }

      final syncedCount = await _syncEngine.syncLatest(
        rows: selectedRows,
        syncRoot: selectedSyncRoot!,
        pspRoot: selectedExternalRoot!,
      );

      await scan();
      statusMessage = syncedCount == 0
          ? "Everything is already in sync."
          : "Synced $syncedCount save folders.";
    } catch (e) {
      statusMessage = "Sync failed: $e";
    } finally {
      isSyncing = false;
      isWorking = false;
      notifyListeners();
    }
  }

  // Deletes a save from PSP storage after confirmation
  Future<bool> deletePSPSave(SaveComparison row) async {
    if (row.psp == null) return false;
    return _deletePaths(
      [Directory(row.psp!.rootUri.toFilePath())],
      status: "Deleted ${row.displayTitle} from PSP storage.",
    );
  }

  // Deletes a save from the sync root after confirmation
  Future<bool> deleteSyncSave(SaveComparison row) async {
    if (row.sync == null) return false;
    return _deletePaths(
      [Directory(row.sync!.rootUri.toFilePath())],
      status: "Deleted ${row.displayTitle} from the sync root.",
    );
  }

  // Deletes a save from both sides
  Future<bool> deleteBothSaves(SaveComparison row) async {
    final paths = <Directory>[];
    if (row.psp != null) paths.add(Directory(row.psp!.rootUri.toFilePath()));
    if (row.sync != null) paths.add(Directory(row.sync!.rootUri.toFilePath()));
    if (paths.isEmpty) return false;
    return _deletePaths(paths, status: "Deleted ${row.displayTitle} from both sides.");
  }

  Future<bool> _deletePaths(List<Directory> dirs, {required String status}) async {
    isWorking = true;
    notifyListeners();
    try {
      for (final dir in dirs) {
        if (await dir.exists()) {
          await dir.delete(recursive: true);
        }
      }
      await scan();
      statusMessage = status;
      return true;
    } catch (e) {
      statusMessage = "Delete failed: $e";
      notifyListeners();
      return false;
    } finally {
      isWorking = false;
      notifyListeners();
    }
  }

  // Imports a new game catalog from a JSON file
  Future<void> importCatalog(File file) async {
    try {
      catalog = await _catalogStore.importCatalog(file);
      await scan();
      statusMessage = "Imported ${catalog.games.length} catalog entries.";
    } catch (e) {
      statusMessage = "Catalog import failed: $e";
    } finally {
      notifyListeners();
    }
  }

  // Restores the selected backup to the sync root
  Future<void> restoreSelectedBackup() async {
    if (selectedSyncRoot == null) return;

    BackupArchive backup;
    try {
      backup = backups.firstWhere((b) => b.id == selectedBackupId);
    } catch (_) {
      statusMessage = "No backup selected.";
      notifyListeners();
      return;
    }

    isWorking = true;
    notifyListeners();

    try {
      if (backupsEnabled) {
        // Create a safety backup before overwriting (preserving the one being restored)
        await _backupStore.createBackup(selectedSyncRoot!, preservedBackup: backup.file);
        await refreshBackups();
      }

      await _backupStore.restoreBackup(backup.file, selectedSyncRoot!);
      await scan();
      await refreshBackups();
      statusMessage = "Restored ${p.basename(backup.file.path)}.";
    } catch (e) {
      statusMessage = "Restore failed: $e";
    } finally {
      isWorking = false;
      notifyListeners();
    }
  }

  // Refreshes the backup list, keeping the selection if still valid
  Future<void> refreshBackups() async {
    try {
      backups = await _backupStore.listBackups();
      if (selectedBackupId == null || !backups.any((b) => b.id == selectedBackupId)) {
        selectedBackupId = backups.isNotEmpty ? backups.first.id : null;
      }
      notifyListeners();
    } catch (e) {
      backups = [];
      selectedBackupId = null;
      statusMessage = "Backups could not be loaded: $e";
      notifyListeners();
    }
  }

  // Row selection helpers
  void toggleSelection(String id) {
    if (selectedRowIDs.contains(id)) {
      selectedRowIDs.remove(id);
    } else {
      selectedRowIDs.add(id);
    }
    notifyListeners();
  }

  void selectAllRows() {
    selectedRowIDs = rows.map((r) => r.id).toSet();
    notifyListeners();
  }

  void clearSelection() {
    selectedRowIDs.clear();
    notifyListeners();
  }

  void setSelectedBackupId(String? id) {
    selectedBackupId = id;
    notifyListeners();
  }

  // Directory pickers
  Future<void> selectExternalRoot() async {
    final path = await FilePicker.getDirectoryPath(
      dialogTitle: "Select PSP storage root",
    );
    if (path != null) {
      selectedExternalRoot = Directory(path);
      await _persistRoot(path, _externalRootKey);
      await scan();
    }
  }

  Future<void> selectSyncRoot() async {
    final path = await FilePicker.getDirectoryPath(
      dialogTitle: "Select sync root",
    );
    if (path != null) {
      selectedSyncRoot = Directory(path);
      await _persistRoot(path, _syncRootKey);
      await scan();
    }
  }

  // Toggle helpers (persist preferences)
  Future<void> toggleBackups(bool value) async {
    backupsEnabled = value;
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool(_backupsEnabledKey, value);
    notifyListeners();
  }

  Future<void> toggleSerialStation(bool value) async {
    useSerialStationAPI = value;
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool(_useSerialStationKey, value);
    notifyListeners();
    if (value) await scan();
  }

  // Fetches SerialStation metadata for saves not already in cache
  Future<Map<String, GameMetadata>> _fetchSerialStationMetadata(List<SaveGame> saves) async {
    final ids = saves.map((s) => s.gameId).toSet()
        .where((id) => GameIDParser.parse(id) != null)
        .toList();

    final result = Map<String, GameMetadata>.from(_serialStationCache);
    final missingIds = ids.where((id) => !result.containsKey(id)).toList();

    if (missingIds.isEmpty) return result;

    statusMessage = "Looking up ${missingIds.length} titles on SerialStation...";
    notifyListeners();

    await Future.wait(missingIds.map((id) async {
      try {
        final meta = await _serialStation.fetchMetadata(id);
        if (meta != null) result[id] = meta;
      } catch (_) {}
    }));

    _serialStationCache = result;
    await _serialStationCache_.save(_serialStationCache);

    return result;
  }

  // Persistence helpers
  Future<void> _persistRoot(String path, String key) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(key, path);
  }

  Future<void> _restoreSelectedRoots() async {
    final prefs = await SharedPreferences.getInstance();
    final extPath = prefs.getString(_externalRootKey);
    final syncPath = prefs.getString(_syncRootKey);
    if (extPath != null) selectedExternalRoot = Directory(extPath);
    if (syncPath != null) selectedSyncRoot = Directory(syncPath);
  }
}
